Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -214,8 +214,17 @@ internal struct BulkWriteOperation<T: Codable>: Operation {
                     mongoc_bulk_operation_execute(bulk, replyPtr, &error)
                 }
 
-                let writeConcern = WriteConcern(from: mongoc_bulk_operation_get_write_concern(bulk))
-                return (serverId, writeConcern.isAcknowledged)
+                var writeConcernAcknowledged: Bool
```

Comment:
As mentioned [here](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#users-cannot-pass-readconcern-or-writeconcern-to-operations-in-transactions), users cannot pass explicit read or write concerns to operations in transactions.

The C driver therefore nulls out the write concern of bulk write operations that are in a transaction. Calling `mongoc_bulk_operation_get_write_concern(bulk)`, however, asserts that the write concern associated with `bulk` is non-null, throwing the following error:
```
mongo-swift-driver/Sources/CLibMongoC/mongoc/mongoc-write-concern.c:132 mongoc_write_concern_journal_is_set(): precondition failed: write_concern'
```
To workaround this, we shouldn't call `mongoc_bulk_operation_get_write_concern(bulk)` for a bulk write operation in a transaction. 

Since we only need the write concern to check whether it is acknowledged or not, it is safe to default to `writeConcernAcknowledged=true` in a transaction because unacknowledged write concerns are [disallowed](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#writeconcern).

Reply:
I think you may have discovered a bug in libmongoc. I don't think `mongoc_bulk_operation_get_write_concern` is intended to ever return NULL. Instead I think it should return an empty write concern (e.g. `mongoc_write_concern_new()`).

So I think the comment should be updated to reflect what you said there, namely that libmongoc returns null write concerns from bulk operations during transactions, so we can only use the returned pointer in other cases. Another related thing, if we've determined the session is in a transaction, we need to throw an error if `self.options.writeConcern != nil` since libmongoc can't verify that for us.

Reply:
Ok, I've updated the comment for now.

Reply:
I've also thrown a `LogicError` if `self.options.writeConcern != nil`. Let me know if you think it should be another error type.

Reply:
if this is a libmongoc bug we should open a CDRIVER ticket

Reply:
To clarify what I meant about updating the comment: bulk write operations in transactions can in fact use write concerns; it's just those write concerns must come from the session, not from the `BulkWriteOptions` passed to the `bulkWrite` helper. The problem we have here is that libmongoc surfaces this implementation detail by nulling out the write concern stored on the bulk write, even if a write concern will be used as part of the transaction. To sidestep this, we can only call `bulk_operation_get_write_concern` if we know we're not in a transaction.

With respect to the error message, this means we should be a bit more specific about what exactly the user did wrong. As it is now, it seems that they can't use a write concern at all, which isn't the case. It could say something like "Cannot specify a write concern on an individual helper in a transaction. Instead specify it when starting the transaction." or something along those lines.

Another thing I just realized is that we're throwing this argument error after we actually executed the bulk write. We should be doing this outside of the `withMongocCollection` block so that the write isn't attempted if we know we'll be throwing an error.

Reply:
One last thing: I think it would make the conditions slightly clearer if we added an `inTransaction` computed property (or something named like that) to `ClientSession` to replace `let transactionState = session?.transactionState, transactionState != .none`.

Reply:
Done.

Reply:
looks good!

I filed [CDRIVER-3607](https://jira.mongodb.org/browse/CDRIVER-3607) for the bug you discovered.